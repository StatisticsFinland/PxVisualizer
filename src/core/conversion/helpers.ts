import { EVariableType, IVariableMeta } from "../types/queryVisualizationResponse";
import { TVariableSelections } from "../types/variableSelections";

type TNullableVariableSelections = TVariableSelections | null | undefined;

export const extractSelectableVariableValues = (selectableVariableCodes: string[], metaData: IVariableMeta[], defaultSelectableVariableCodes: TNullableVariableSelections, selectedValueCodes: TNullableVariableSelections) => {
    if (selectableVariableCodes.length === 0) {
        return {};
    }

    let selectableVariableCodesCheckList = selectableVariableCodes;
    let currentSelectableVariableValues: { [key: string]: string[] } = {};

    if (selectedValueCodes) {
        const { parsedValues, updatedArray } = parseSelected(selectableVariableCodesCheckList, selectedValueCodes);
        selectableVariableCodesCheckList = updatedArray;
        currentSelectableVariableValues = {...currentSelectableVariableValues, ...parsedValues};
        if (selectableVariableCodesCheckList.length === 0) {
            return currentSelectableVariableValues;
        }
    }

    if (defaultSelectableVariableCodes) {
        const { parsedValues, updatedArray } = parseDefault(selectableVariableCodesCheckList, defaultSelectableVariableCodes);
        selectableVariableCodesCheckList = updatedArray;
        currentSelectableVariableValues = {...currentSelectableVariableValues, ...parsedValues};
        if (selectableVariableCodesCheckList.length === 0) {
            return currentSelectableVariableValues;
        }
    }

    selectableVariableCodesCheckList.forEach((variableCode) => {
        const variableIndex = metaData.findIndex(metaDataObject => metaDataObject.code === variableCode);
        if (variableIndex > -1) {
            if (metaData[variableIndex].type === EVariableType.Time) {
                currentSelectableVariableValues[variableCode] = [metaData[variableIndex].values[metaData[variableIndex].values.length - 1].code];
            } else {
                currentSelectableVariableValues[variableCode] = [metaData[variableIndex].values[0].code];
            }
        }
    });

    return currentSelectableVariableValues;
}

const parseSelected = (selectableVariableCodesCheckList: string[], selectedValueCodes: TVariableSelections) => {
    const selectedVariables = Object.keys(selectedValueCodes);
    const parsedValues: { [key: string]: string[] } = {};
    const updatedArray = selectableVariableCodesCheckList.map((code) => {
        const variableIndex = selectedVariables.findIndex(variableCode => variableCode === code);
        if (variableIndex > -1) {
            if (selectedValueCodes[code].length > 0) {
                parsedValues[code] = selectedValueCodes[code];
                return '';
            } else {
                return code;
            }
        } else {
            return code;
        }
    }).filter(element => element.length > 0);

    return {parsedValues, updatedArray};
};

const parseDefault = (selectableVariableCodesCheckList: string[], defaultSelectableVariableCodes: TVariableSelections) => {
    const defaultVariables = Object.keys(defaultSelectableVariableCodes);
    const parsedValues: { [key: string]: string[] } = {};
    const updatedArray = selectableVariableCodesCheckList.map((code) => {
            const variableIndex = defaultVariables.findIndex(variableCode => variableCode === code);
            if (variableIndex > -1) {
                parsedValues[code] = defaultSelectableVariableCodes[defaultVariables[variableIndex]];
                return '';
            } else {
                return code;
            }
        }).filter(element => element.length > 0);

    return {parsedValues, updatedArray};
};