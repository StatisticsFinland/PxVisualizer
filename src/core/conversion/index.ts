export { convertPxGraphDataToChartOptions } from './pxGrafDataConverter';
export { extractSelectableVariableValues } from './helpers';
export { convertPxGrafResponseToView } from './viewUtils';
export { TTranslationPackage, AddTranslations } from './translations';