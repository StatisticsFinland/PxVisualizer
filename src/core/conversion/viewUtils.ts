import { IQueryVisualizationResponse } from "../types";
import { EVariableType, EVisualizationType, IContentComponent, IVariableMeta, IVariableValueMeta, TMultiLanguageString } from "../types/queryVisualizationResponse";
import { ESeriesType, IDataCell, IDataSeries, IUnitInfo, TValueSelectionAmounts, View } from "../types/view";
import { ASCENDING, ASCENDING_SORTING_FUNC, DESCENDING, DESCENDING_SORTING_FUNC, NO_SORTING, SUM, REVERSED, sortViewBasedOnSeries, sortViewBasedOnSeriesRelative, sortViewBasedOnSum, reverseViewOrder } from "./viewSorting";
import { cartesianProduct, onlyUnique } from "./utilityFunctions";
import { TVariableSelections } from "../types/variableSelections";
import Translations from "./translations";

export function convertPxGrafResponseToView(
    responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections
): View {
    const view: View = convert(responseObj, selectedValueCodes, getValueSelectAmounts(selectedValueCodes));

    const isRelativeChart =
        view.visualizationSettings?.visualizationType === EVisualizationType.PercentHorizontalBarChart
        || view.visualizationSettings?.visualizationType === EVisualizationType.PercentVerticalBarChart;

    switch (responseObj.visualizationSettings?.sorting) {
        case null: return view;
        case undefined: return view;
        case NO_SORTING: return view;
        case REVERSED: return reverseViewOrder(view);
        case SUM: return sortViewBasedOnSum(view, DESCENDING_SORTING_FUNC);
        case ASCENDING: return sortViewBasedOnSeries(view, 0, ASCENDING_SORTING_FUNC);
        case DESCENDING: return sortViewBasedOnSeries(view, 0, DESCENDING_SORTING_FUNC);
        default:
            if (isRelativeChart) return sortViewBasedOnSeriesRelative(view, convertToRelative(view), getSortingIndex(responseObj), DESCENDING_SORTING_FUNC);
            else return sortViewBasedOnSeries(view, getSortingIndex(responseObj), DESCENDING_SORTING_FUNC);
    };
}

export function convertToRelative(input: View): View {
    const newSeries = input.series.map(s => ({
        ...s,
        series: s.series.map((dataCell, index) => {
            const sum = input.series.reduce((sum, series) => sum + (series.series[index].value ?? 0), 0);
            const value = (!sum || sum === 0) ? null : Math.round((100 * (dataCell.value ?? 0) / sum) * 10) / 10;
            const newCell: IDataCell = { value, precision: 1, preliminary: dataCell.preliminary };
            if (value === null) newCell.missingCode = dataCell.missingCode;
            return newCell;
        }),
    }));

    const convertUnit = (unit: TMultiLanguageString) => {
        return Object.keys(unit).reduce((acc: TMultiLanguageString, lang: string) => {
            acc[lang] = '%';
            return acc;
        }, {})
    };

    return {...input, units: input.units.map(u => ({name: u.name, unit: convertUnit(u.unit) })) , series: newSeries};
}

function getValueSelectAmounts(selectedValueCodes: TVariableSelections): TValueSelectionAmounts {
    const result: TValueSelectionAmounts = {};
    Object.entries(selectedValueCodes).forEach(([key, value]) => result[key] = value.length);
    return result;
}

function convert(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections, selectedValueAmounts: TValueSelectionAmounts): View {
    const { metaData, selectableVariableCodes } = responseObj;
    const selectableVariables = getSelectableVariables(metaData, selectableVariableCodes);
    const contentVar = getContentVariable(metaData);
    const unsortedSeries = buildSeries(responseObj, selectedValueCodes);

    return {
        header: responseObj.header,
        tableReferenceName: responseObj.tableReference.name,
        subheaderValues: getSubheaderValues(selectableVariables, selectedValueCodes, selectedValueAmounts),
        units: getUnitInformation(contentVar, selectedValueCodes),
        sources: getContentProperty(contentVar, selectedValueCodes, (cc) => cc?.source ?? Translations.empty),
        columnNameGroups: unsortedSeries.columnNameGroups,
        series: unsortedSeries.series,
        colVarNames: getVariableNames(responseObj.columnVariableCodes, responseObj.metaData),
        rowVarNames: getVariableNames(responseObj.rowVariableCodes, responseObj.metaData),
        selectableVarNames: getVariableNames(responseObj.selectableVariableCodes, responseObj.metaData),
        visualizationSettings: responseObj.visualizationSettings,
        seriesType: getSeriesType(responseObj.columnVariableCodes, responseObj.metaData)
    };
}

export function buildSeries(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections): { columnNameGroups: TMultiLanguageString[][], series: IDataSeries[] } {
    const valuesInView: Set<string> = new Set(getValuesInView(responseObj, selectedValueCodes).map(v => v.code));
    const rowVarValues: IVariableValueMeta[][] = getVariableValues(responseObj, responseObj.rowVariableCodes);
    const columnVarValues: IVariableValueMeta[][] = getVariableValues(responseObj, responseObj.columnVariableCodes);
    const selectableValues: IVariableValueMeta[][] = getVariableValues(responseObj, responseObj.selectableVariableCodes);
    const cartesianColumnVarValues: IVariableValueMeta[][] = cartesianProduct(columnVarValues);
    const cartesianRowVarValues: IVariableValueMeta[][] = cartesianProduct(rowVarValues);
    const cartesianSelectableVarValues: IVariableValueMeta[][] = cartesianProduct(selectableValues);
    const viewPrecision: number | null = getViewPrecision(responseObj, selectedValueCodes);
    const multiselectValues: IVariableValueMeta[] = responseObj.metaData.find(v => v.code === responseObj.visualizationSettings.multiselectableVariableCode)?.values ?? [];

    let dataIndex: number = 0;
    const viewSeries: IDataSeries[] = [];

    cartesianSelectableVarValues.forEach(selectableVarValueGroup => {
        const multiselects = selectableVarValueGroup.filter(v => multiselectValues.includes(v));
        cartesianRowVarValues.forEach(rowVarValueGroup => {
            const preliminaryRow: boolean = rowVarValueGroup.some(v => Object.values(v.name)[0].trim().endsWith('*'));
            const rowPrecision: number | null = viewPrecision ?? rowVarValueGroup.find(v => v.contentComponent)?.contentComponent?.numberOfDecimals ?? null;
            const rowSeries: IDataCell[] = [];

            cartesianColumnVarValues.forEach(colVarValueGroup => {
                const combinedValues: IVariableValueMeta[] = [...rowVarValueGroup, ...colVarValueGroup, ...selectableVarValueGroup];
                const dataCell: IDataCell | null = createDataCell(
                    combinedValues,
                    valuesInView,
                    responseObj,
                    dataIndex,
                    rowPrecision,
                    preliminaryRow,
                    colVarValueGroup);
                if (dataCell) {
                    rowSeries.push(dataCell);
                }
                dataIndex++;
            });

            if (rowSeries.length > 0) {
                viewSeries.push({
                    rowNameGroup: [...multiselects.map(value => value.name), ...rowVarValueGroup.map(value => value.name)],
                    series: rowSeries
                });
            }
        });
    });

    return {
        columnNameGroups: cartesianColumnVarValues.map(columnVarValueGroup => columnVarValueGroup.map(value => value.name)),
        series: viewSeries
    };
}

function createDataCell(
    combinedValues: IVariableValueMeta[],
    valuesInView: Set<string>,
    responseObj: IQueryVisualizationResponse,
    dataIndex: number,
    rowPrecision: number | null,
    preliminaryRow: boolean,
    colVarValueGroup: IVariableValueMeta[]
): IDataCell | null {
    if (combinedValues.every(value => valuesInView.has(value.code))) {
        const dataCell: IDataCell = {
            value: responseObj.data[dataIndex],
            precision: rowPrecision ?? colVarValueGroup.find(v => v.contentComponent)?.contentComponent?.numberOfDecimals ?? 0,
            preliminary: preliminaryRow || colVarValueGroup.some(v => Object.values(v.name)[0].trim().endsWith('*'))
        };

        if (!dataCell.value) dataCell.missingCode = responseObj.missingDataInfo[dataIndex];
        return dataCell;
    }
    else {
        return null;
    }
}

function getValuesInView(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections): IVariableValueMeta[] {
    return responseObj.metaData.flatMap(variable => {
        if (responseObj.selectableVariableCodes.includes(variable.code) ||
            Object.keys(selectedValueCodes).includes(variable.code)) {
            const values: IVariableValueMeta[] = variable.values.filter(value => selectedValueCodes[variable.code].includes(value.code));
            if (values.length === 0) throw new Error("Provided selected value code can not be found from the metadata");
            return values;
        }
        else {
            return variable.values;
        }
    });
}

function getVariableValues(responseObj: IQueryVisualizationResponse, variableCodes: string[]): IVariableValueMeta[][] {
    return responseObj.metaData
        .filter(vm => variableCodes.includes(vm.code))
        .filter(vm => vm.values.length > 1)
        .map(vm => vm.values);
}

function getViewPrecision(responseObj: IQueryVisualizationResponse, selectedValueCodes: TVariableSelections): number | null {
    const contentVar = getContentVariable(responseObj.metaData);

    // Only one content variable value
    if (contentVar.values.length === 1) {
        return contentVar.values[0].contentComponent?.numberOfDecimals ?? null

    // Content variable in selection
    } else if (responseObj.selectableVariableCodes.includes(contentVar.code) && responseObj.visualizationSettings.multiselectableVariableCode !== contentVar.code) {
        return contentVar.values.find(value => selectedValueCodes[contentVar.code].includes(value.code))?.contentComponent?.numberOfDecimals ?? null;

    } else {
        return null;
    }
}

function getVariableNames(varCodes: string[], meta: IVariableMeta[]): TMultiLanguageString[] {
    return varCodes.reduce((acc: TMultiLanguageString[], code: string) => {
        const name = meta.find(v => v.code === code)?.name;
        if (name) acc.push(name);
        return acc;
    }, [])
}

function getSelectableVariables(metaData: IVariableMeta[], selectableVariableCodes: string[]): IVariableMeta[] {
    return metaData.filter(varMeta => selectableVariableCodes.includes(varMeta.code));
}

function getSortingIndex(responseObj: IQueryVisualizationResponse): number {
    const sortingVar = responseObj.metaData.find(v => v.code == responseObj.rowVariableCodes[0]);
    const valIndex = sortingVar?.values.findIndex(v => v.code === responseObj.visualizationSettings?.sorting);
    return (!valIndex || valIndex < 0) ? 0 : valIndex;
}

function getContentVariable(varMeta: IVariableMeta[]): IVariableMeta {
    const contentVar = varMeta.find(v => v.type === EVariableType.Content);
    if (contentVar) return contentVar;
    else throw new Error('Content variable is not defined');
}

function getSubheaderValues(selectableVariables: IVariableMeta[], selectedValueCodes: TVariableSelections, selectedValueAmounts: TValueSelectionAmounts): TMultiLanguageString[] {
    const subheaderValues: TMultiLanguageString[] = [];
    selectableVariables.forEach(selectableVariable => {
        // We only want to show the value name if there is exactly one selected.
        if (selectedValueAmounts[selectableVariable.code] == 1) {
            // Find the value whose code matches selected value code for this variable and add it to subheaderValues
            subheaderValues.push(selectableVariable.values.find(v => v.code === selectedValueCodes[selectableVariable.code][0])?.name ?? {})
        }
    });
    return subheaderValues;
}

function getUnitInformation(contentVar: IVariableMeta, selectedValueCodes: TVariableSelections): IUnitInfo[] {
    const values = contentVar.code in selectedValueCodes
        ? contentVar.values.filter(v => selectedValueCodes[contentVar.code].includes(v.code))
        : contentVar.values

    return values.reduce((accum: IUnitInfo[], valMeta: IVariableValueMeta) => {
        if(valMeta.contentComponent) accum.push({
            name: valMeta.name,
            unit: valMeta.contentComponent.unit
        });
        return accum;
    }, []);
}

function getContentProperty(
    contentVar: IVariableMeta,
    selectedValueCodes: TVariableSelections,
    extractorFunc: (cc: IContentComponent | null) => TMultiLanguageString
): TMultiLanguageString[] {
    if (contentVar.code in selectedValueCodes) {
        return contentVar.values
            .filter(v => selectedValueCodes[contentVar.code].includes(v.code))
            .map(cvv => extractorFunc(cvv.contentComponent))
            .filter(onlyUnique);
    } else {
        return contentVar.values.map(cvv => extractorFunc(cvv.contentComponent)).filter(onlyUnique);
    }
}

function getSeriesType (varCodes: string[], meta: IVariableMeta[]) {
    if (varCodes.length > 1 || varCodes.length === 0) {
        return ESeriesType.Nominal;
    }

    const metaDataType = meta.find(v => v.code === varCodes[0])?.type;

    switch (metaDataType) {
        case EVariableType.Time:
            return ESeriesType.Time;
        case EVariableType.Ordinal:
            return ESeriesType.Ordinal;
        default:
            return ESeriesType.Nominal;
    }
}