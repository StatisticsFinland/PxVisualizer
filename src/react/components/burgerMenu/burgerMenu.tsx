import React from "react";
import styled from "styled-components";
import MenuItem from "./menuItem/menuItem";
import HighchartsReact from "highcharts-react-official";
import { View } from "../../../core/types/view";
import Translations from "../../../core/conversion/translations";
import Icon from "../icon/icon";
import { viewToDownloadCSVOption } from "../../../core/tables/csvTable";
import { generateFilename } from "../../../core/tables/exportingUtils";
import { TIcon } from "../../../core/types/icon";
import { viewToDownloadXLSOption } from "../../../core/tables/xlsx/xlsxBuilder";

const BurgerWrapper = styled.div`
    width: 2.5rem;
    height: 2.5rem;
`;

const Hamburger = styled.button`
    width: 100%;
    height: 100%;
    svg {
        height: 1.5rem;
        width: 1.5rem;
    }
    background-color: transparent;
    border: none;
    cursor: pointer;
`;

const MenuAnchor = styled.div`
    width: 100%;
    height: 0;
    position: relative;
`;

const MenuWrapper = styled.div`
    display: flex;
    flex-flow: column;
    position: absolute;
    background-color: #fff;
    border-radius: 18px;
    top: 0.2rem;
    right: 0.2rem;
    border: 1px solid #bdbdbd;
`
const List = styled.ul`
    list-style-type: none;
    margin: 0;
    padding: 0;
`;

export interface IBurgerMenuProps {
    viewData: View;
    currentChartRef?: HighchartsReact.RefObject | null;
    locale: string;
    tableToggle?: {
        toggleHandler: () => void;
        tableMode: boolean;
    };
    menuItemDefinitions?: (IFunctionalMenuItem | ILinkMenuItem)[];
    menuIconInheritColor?: boolean;
    exportWidth?: number;
    exportHeight?: number,
}

interface IBaseMenuItem {
    text: string;
    prefixIcon?: TIcon | Exclude<React.ReactNode, string>;
    suffixIcon?: TIcon | Exclude<React.ReactNode, string>;
}

export interface IFunctionalMenuItem extends IBaseMenuItem {
    onClick: () => void;
}

export interface ILinkMenuItem extends IBaseMenuItem {
    url: string;
    isExternal?: boolean;
    openNewTab?: boolean;
}

export function calculateExportDimensions(inputWidth: number | undefined, inputHeight: number | undefined): { finalWidth: number, finalHeight: number } {
    const MIN_EXPORT_ASPECT_RATIO: number = 0.5;
    const MAX_EXPORT_ASPECT_RATIO: number = 2;
    const MIN_EXPORT_DIMENSION: number = 600;
    const MAX_EXPORT_DIMENSION: number = 3000;

    inputWidth = inputWidth ?? MIN_EXPORT_DIMENSION;
    inputHeight = inputHeight ?? MIN_EXPORT_DIMENSION;

    // Aspect ratio is the ratio of width to height clamped between MIN_EXPORT_ASPECT_RATIO and MAX_EXPORT_ASPECT_RATIO.
    let aspectRatio: number = Math.max(
        MIN_EXPORT_ASPECT_RATIO,
        Math.min(
            MAX_EXPORT_ASPECT_RATIO,
            inputWidth / inputHeight
        )
    );

    let finalWidth: number;
    let finalHeight: number;

    // If aspect ratio is less than 1, then height is greater than width.
    if (aspectRatio < 1) {
        finalHeight = Math.max(MIN_EXPORT_DIMENSION, Math.min(MAX_EXPORT_DIMENSION, inputHeight));
        finalWidth = finalHeight * aspectRatio;
        if (finalWidth < MIN_EXPORT_DIMENSION) {
            finalWidth = MIN_EXPORT_DIMENSION;
            finalHeight = finalWidth / aspectRatio;
        }
    } else {
        finalWidth = Math.max(MIN_EXPORT_DIMENSION, Math.min(MAX_EXPORT_DIMENSION, inputWidth));
        finalHeight = finalWidth / aspectRatio;
        if (finalHeight < MIN_EXPORT_DIMENSION) {
            finalHeight = MIN_EXPORT_DIMENSION;
            finalWidth = finalHeight * aspectRatio;
        }
    }

    return { finalWidth, finalHeight };
}

export const BurgerMenu: React.FC<IBurgerMenuProps> = ({viewData, currentChartRef, tableToggle, menuItemDefinitions, locale, menuIconInheritColor = false, exportWidth, exportHeight}) => {
    const [isOpen, setIsOpen] = React.useState(false);

    const menuRef = React.useRef<any>(null);
    const closeMenu = (e: Event) => {
        if(menuRef.current && isOpen && !menuRef.current.contains(e.target)) {
            setIsOpen(false);
          }
    }

    document.addEventListener('mousedown', closeMenu);
    const exportCSV = viewToDownloadCSVOption(viewData, locale);
    const exportXLSX = viewToDownloadXLSOption(viewData, locale);
    const customMenuItemArray = menuItemDefinitions?.map((menuItemDefinition, index) => {
        if ('onClick' in menuItemDefinition) {
            return <MenuItem isFirst={index === 0} bottomSeparator={index + 1 === menuItemDefinitions.length} locale={locale} prefixIcon={menuItemDefinition.prefixIcon} suffixIcon={menuItemDefinition.suffixIcon} key={`customfmenuitem-${index}`} text={menuItemDefinition.text} onClick={menuItemDefinition.onClick} />; // NOSONAR
        }

        if ('url' in menuItemDefinition) {
            return <MenuItem isFirst={index === 0} bottomSeparator={index + 1 === menuItemDefinitions.length} locale={locale} prefixIcon={menuItemDefinition.prefixIcon} suffixIcon={menuItemDefinition.suffixIcon} isExternal={menuItemDefinition.isExternal} key={`customlmenuitem-${index}`} text={`${menuItemDefinition.text}`} url={menuItemDefinition.url} openNewTab={menuItemDefinition.openNewTab} />; // NOSONAR
        }
    });

    return (
        <BurgerWrapper ref={menuRef}>
            <Hamburger aria-label={`${Translations.chartMenuLabel[locale]}`} aria-expanded={isOpen} onClick={() => {setIsOpen(!isOpen)}}>
                <Icon inheritColor={menuIconInheritColor} icon={isOpen ? 'Times' : 'Bars'} />
            </Hamburger>
            <MenuAnchor>
                {
                    isOpen &&
                    <MenuWrapper>
                        <List>
                            {menuItemDefinitions && customMenuItemArray}
                            <MenuItem isFirst={!menuItemDefinitions} locale={locale} prefixIcon={'Download'} text={exportXLSX.text} onClick={() => exportXLSX.onClick()} />
                            <MenuItem isLast={!currentChartRef} bottomSeparator={!!currentChartRef} locale={locale} prefixIcon={'Download'} text={exportCSV.text} onClick={() => exportCSV.onClick()} />
                            {
                                    currentChartRef &&
                                    <>
                                        <MenuItem
                                            locale={locale}
                                            prefixIcon={'Download'}
                                            text={Translations.downloadSVG[locale]}
                                            onClick={() =>
                                                currentChartRef.chart.exportChartLocal({
                                                    filename: `${generateFilename(viewData.tableReferenceName)}`,
                                                    type: "image/svg+xml",
                                                    sourceWidth: calculateExportDimensions(currentChartRef.chart.chartWidth, currentChartRef.chart.chartHeight).finalWidth,
                                                    sourceHeight: calculateExportDimensions(currentChartRef.chart.chartWidth, currentChartRef.chart.chartHeight).finalHeight,
                                                    scale: 1
                                                }, {})
                                            } />
                                        <MenuItem
                                            isLast={!tableToggle}
                                            bottomSeparator={!!tableToggle}
                                            locale={locale}
                                            prefixIcon={'Download'}
                                            text={Translations.downloadPNG[locale]}
                                            onClick={() =>
                                                currentChartRef.chart.exportChartLocal({
                                                    filename: `${generateFilename(viewData.tableReferenceName)}`,
                                                    sourceWidth: calculateExportDimensions(currentChartRef.chart.chartWidth, currentChartRef.chart.chartHeight).finalWidth,
                                                    sourceHeight: calculateExportDimensions(currentChartRef.chart.chartWidth, currentChartRef.chart.chartHeight).finalHeight,
                                                    scale: 1
                                                }, {})
                                            } />
                                    </>
                                }
                                {
                                    tableToggle &&
                                    <MenuItem isLast={true} locale={locale} text={tableToggle.tableMode ? Translations.toggleTableModeOffText[locale] : Translations.toggleTableModeOnText[locale]} onClick={() => tableToggle.toggleHandler()} />
                                }
                            </List>
                    </MenuWrapper>
                }
            </MenuAnchor>
        </BurgerWrapper>
    );
}

export default BurgerMenu;