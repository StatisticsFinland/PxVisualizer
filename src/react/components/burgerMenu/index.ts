export * from './burgerMenu';
export * from './menuItem';