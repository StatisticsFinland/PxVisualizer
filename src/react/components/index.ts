export * from "./chart";
export * from "./ErrorBoundary";
export * from "./globalStyle";
export * from "./icon";
export * from "./burgerMenu";
