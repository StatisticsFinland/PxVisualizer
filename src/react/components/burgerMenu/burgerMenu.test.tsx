import { act, render, screen, waitFor } from "@testing-library/react";
import { BurgerMenu, calculateExportDimensions } from "./burgerMenu";
import { HORIZONTAL_BAR_CHART_ASCENDING } from "../../../core/conversion/TestFixtures/horizontalBarChart";
import { convertPxGrafResponseToView } from "../../../core/conversion/viewUtils";
import React from "react";
import 'jest-styled-components';

describe('burgerMenu, rendering tests', () => {
    it('Should render', () => {
        const view = convertPxGrafResponseToView(HORIZONTAL_BAR_CHART_ASCENDING, {});
        const { asFragment } = render(<BurgerMenu locale="fi" viewData={view} />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('burgerMenu, functional tests', () => {
    it('Should open the menu with correct menu elements', async() => {
        const view = convertPxGrafResponseToView(HORIZONTAL_BAR_CHART_ASCENDING, {});
        render(<BurgerMenu locale="fi" viewData={view} />);
        act(() => {
            screen.getByRole('button').click();
        });
        await waitFor(() => {
            expect(screen.getAllByRole('button').length).toEqual(3);
            expect(screen.getByText('Lataa taulukko (csv)')).toBeTruthy();
            expect(screen.getByText('Lataa taulukko (xlsx)')).toBeTruthy();
        });
    });

    it('Should open the menu with correct custom elements', async() => {
        const view = convertPxGrafResponseToView(HORIZONTAL_BAR_CHART_ASCENDING, {});
        render(<BurgerMenu locale="fi" viewData={view} menuItemDefinitions={[{text: 'Foo', onClick: jest.fn()}, {text: 'Bar', onClick: jest.fn()}, {text: 'Baz', url: 'foobar.com', isExternal: true}, {text: 'Baz2', url: 'foobar2.com', isExternal: false}]} />);
        act(() => {
            screen.getByRole('button').click();
        });
        await waitFor(() => {
            expect(screen.getAllByRole('button').length).toEqual(7);
            expect(screen.getByText('Lataa taulukko (csv)')).toBeTruthy();
            expect(screen.getByText('Lataa taulukko (xlsx)')).toBeTruthy();
            expect(screen.getByText('Foo')).toBeTruthy();
            expect(screen.getByText('Bar')).toBeTruthy();
            expect(screen.getByText('Baz')).toBeTruthy();
            expect(screen.getByText('Ulkoinen linkki')).toBeTruthy();
            expect(screen.getByText('Baz2')).toBeTruthy();
            expect(screen.getAllByRole('link').length).toBe(2);
        });
    });

    it('Should invoke the custom function when clicked', async() => {
        const view = convertPxGrafResponseToView(HORIZONTAL_BAR_CHART_ASCENDING, {});
        const mockFunction = jest.fn();
        const btnText = 'Foo';
        render(<BurgerMenu locale="fi" viewData={view} menuItemDefinitions={[{text: btnText, onClick: mockFunction}]} />);
        act(() => {
            screen.getByRole('button').click();
        });
        await waitFor(() => {
            screen.getByText(btnText).click();
            expect(mockFunction).toBeCalledTimes(1);
        });
    });

    it('Should calculate correct export dimensions for compliant input size', () => {
        const { finalWidth, finalHeight } = calculateExportDimensions(1200, 800);
        expect(finalWidth).toEqual(1200);
        expect(finalHeight).toEqual(800);
    });

    it('Should calculate correct export dimensions when aspect ratio is too low', () => {
        const { finalWidth, finalHeight } = calculateExportDimensions(2000, 600);
        expect(finalWidth).toEqual(2000);
        expect(finalHeight).toEqual(1000);
    });

    it('Should calculate correct export dimensions when aspect ratio is too large', () => {
        const { finalWidth, finalHeight } = calculateExportDimensions(600, 2000);
        expect(finalWidth).toEqual(1000);
        expect(finalHeight).toEqual(2000);
    });

    it('Should calculate correct export dimensions when width is too low', () => {
        const { finalWidth, finalHeight } = calculateExportDimensions(400, 800);
        expect(finalWidth).toEqual(600);
        expect(finalHeight).toEqual(1200);
    });

    it('Should calculate correct export dimensions when width is too large', () => {
        const { finalWidth, finalHeight } = calculateExportDimensions(4000, 2000);
        expect(finalWidth).toEqual(3000);
        expect(finalHeight).toEqual(1500);
    });

    it('Should calculate correct export dimensions when height is too low', () => {
        const { finalWidth, finalHeight } = calculateExportDimensions(800, 400);
        expect(finalWidth).toEqual(1200);
        expect(finalHeight).toEqual(600);
    });

    it('Should calculate correct export dimensions when height is too large', () => {
        const { finalWidth, finalHeight } = calculateExportDimensions(2000, 4000);
        expect(finalWidth).toEqual(1500);
        expect(finalHeight).toEqual(3000);
    });

    it('Should calculate correct export dimensions width and height are not provided', () => {
        const { finalWidth, finalHeight } = calculateExportDimensions(undefined, undefined);
        expect(finalWidth).toEqual(600);
        expect(finalHeight).toEqual(600);
    });
});