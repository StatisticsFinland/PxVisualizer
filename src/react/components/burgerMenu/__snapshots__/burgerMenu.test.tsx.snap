// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`burgerMenu, rendering tests Should render 1`] = `
<DocumentFragment>
  .c0 {
  width: 2.5rem;
  height: 2.5rem;
}

.c1 {
  width: 100%;
  height: 100%;
  background-color: transparent;
  border: none;
  cursor: pointer;
}

.c1 svg {
  height: 1.5rem;
  width: 1.5rem;
}

.c2 {
  width: 100%;
  height: 0;
  position: relative;
}

<div
    class="c0"
  >
    <button
      aria-expanded="false"
      aria-label="Kuvion valikko"
      class="c1"
    >
      <svg
        aria-hidden="true"
        height="29.92mm"
        viewBox="0 0 113.12 84.82"
        width="39.91mm"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m56.51,14.13c-16.33,0-32.67.01-49,0-5.37,0-8.89-4.83-6.99-9.59C1.63,1.76,3.77.26,6.76.01c.22-.02.44,0,.66,0,32.74,0,65.49,0,98.23,0,4.3,0,7.45,2.99,7.47,7.01.02,4.11-3.15,7.1-7.6,7.11-16.33.01-32.67,0-49,0Z"
          style="fill: #1A3061;"
        />
        <path
          d="m56.55,49.47c-16.41,0-32.82.01-49.22,0-3.78,0-6.77-2.56-7.26-6.09-.49-3.53,1.78-6.9,5.35-7.81.77-.2,1.6-.21,2.4-.21,32.52,0,65.04-.01,97.57,0,4.51,0,7.69,2.89,7.74,6.97.04,4.19-3.16,7.14-7.79,7.15-16.26,0-32.52,0-48.78,0Z"
          style="fill: #1A3061;"
        />
        <path
          d="m56.43,84.82c-16.33,0-32.67,0-49,0-3.68,0-6.45-2.17-7.25-5.56-.73-3.09.92-6.54,3.92-7.84,1.04-.45,2.27-.69,3.41-.69,32.74-.04,65.49-.03,98.23-.02,4.27,0,7.41,3.06,7.39,7.09-.02,4.02-3.17,7.02-7.47,7.02-16.41.02-32.82,0-49.22,0Z"
          style="fill: #1A3061;"
        />
      </svg>
    </button>
    <div
      class="c2"
    />
  </div>
</DocumentFragment>
`;
